var/spack/repos/builtin/packages/libzmq/package.py

```diff
@@ -5,120 +5,131 @@
     version("4.3.4", sha256="c593001a89f5a85dd2ddf564805deb860e02471171b3f204944857336295c3e5")
     version("4.3.3", sha256="9d9285db37ae942ed0780c016da87060497877af45094ff9e1a1ca736e3875a2")
-    version("4.3.2", sha256="ebd7b5c830d6428956b67a0454a7f8cbed1de74b3b01e5c33c5378e22740f763")
-    version("4.3.1", sha256="bcbabe1e2c7d0eec4ed612e10b94b112dd5f06fcefa994a0c79a45d835cd21eb")
-    version("4.3.0", sha256="8e9c3af6dc5a8540b356697081303be392ade3f014615028b3c896d0148397fd")
-    version("4.2.5", sha256="cc9090ba35713d59bb2f7d7965f877036c49c5558ea0c290b0dcc6f2a17e489f")
-    version("4.2.2", sha256="5b23f4ca9ef545d5bd3af55d305765e3ee06b986263b31967435d285a3e6df6b")
-    version("4.1.4", sha256="e99f44fde25c2e4cb84ce440f87ca7d3fe3271c2b8cfbc67d55e4de25e6fe378")
-    version("4.1.2", sha256="f9162ead6d68521e5154d871bac304f88857308bb02366b81bb588497a345927")
-    version("4.1.1", sha256="43d61e5706b43946aad4a661400627bcde9c63cc25816d4749c67b64c3dab8db")
-    version("4.0.7", sha256="e00b2967e074990d0538361cc79084a0a92892df2c6e7585da34e4c61ee47b03")
-    version("4.0.6", sha256="28a2a9c9b77014c39087a498942449df18bb9885cdb63334833525a1d19f2894")
-    version("4.0.5", sha256="3bc93c5f67370341428364ce007d448f4bb58a0eaabd0a60697d8086bc43342b")
-
-    variant(
-        "libsodium",
-        default=True,
-        description="Build with message encryption support via libsodium",
-    )
-
-    variant("drafts", default=False, description="Build and install draft classes and methods")
 
+    # Policy: https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions
+    #
+    # Version   Deprecation Reason
+    # =======   ===============================================================================
+    # @:4.3.2   https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-13132,
+    #           https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15166,
+    #           https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-36400,
+    #           https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-20234,
+    #           https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-20235,
+    #           https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-20236,
+    #           https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-20237,
+    #           poor modern compiler support, see conflicts
+    # @:4.3.1   https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-6250
+    #   @:4.2   legacy release 2018 (https://github.com/zeromq/libzmq/releases/tag/v4.2.5)
+    #   @:4.1   legacy release 2016 (http://wiki.zeromq.org/intro:get-the-software),
+    #           http://download.zeromq.org/zeromq-4.1.4.tar.gz -> 503 Service Unavailable
+    #   @:4.0   legacy release 2016 (http://wiki.zeromq.org/intro:get-the-software),
+    #           http://download.zeromq.org/zeromq-4.0.7.tar.gz -> 503 Service Unavailable
+
+    def deprecated(*values, **kwargs):
```

Comment:
can you just use version(..., deprecated=True) please?

Reply:
How do you feel about adding it as an official DSL helper? e.g.

```diff
--- a/lib/spack/spack/directives.py
+++ b/lib/spack/spack/directives.py
@@ -25,7 +25,7 @@ class OpenMpi(Package):
   * ``provides``
   * ``resource``
   * ``variant``
-  * ``version``
+  * ``version`` (``deprecated``)
   * ``requires``

 """
@@ -70,6 +70,7 @@ class OpenMpi(Package):
     "resource",
     "build_system",
     "requires",
+    "deprecated",
 ]

 #: These are variant names used by Spack internally; packages can't use them
@@ -77,7 +78,7 @@ class OpenMpi(Package):

 #: Names of possible directives. This list is mostly populated using the @directive decorator.
 #: Some directives leverage others and in that case are not automatically added.
-directive_names = ["build_system"]
+directive_names = ["build_system", "deprecated"]

 _patch_order_index = 0

@@ -490,6 +491,15 @@ def _depends_on(pkg, spec, when=None, type=dt.DEFAULT_TYPES, patches=None):
         execute_patch(dependency)


+def deprecated(*values, **kwargs):
+    """Helper to declare a deprecated version
+
+    which fits into a single line for the standard case
+    """
+    kwargs["deprecated"] = True
+    return version(*values, **kwargs)
+
+
 @directive("conflicts")
 def conflicts(conflict_spec, when=None, msg=None):
     """Allows a package to define a conflict.
--- a/var/spack/repos/builtin/packages/libzmq/package.py
+++ b/var/spack/repos/builtin/packages/libzmq/package.py
@@ -46,10 +46,6 @@ def url_for_version(self, ver):
     #   @:4.0   legacy release 2016 (http://wiki.zeromq.org/intro:get-the-software),
     #           http://download.zeromq.org/zeromq-4.0.7.tar.gz -> 503 Service Unavailable

-    def deprecated(*values, **kwargs):
-        kwargs["deprecated"] = True
-        return version(*values, **kwargs)
-
     deprecated("4.3.2", sha256="ebd7b5c830d6428956b67a0454a7f8cbed1de74b3b01e5c33c5378e22740f763")
     deprecated("4.3.1", sha256="bcbabe1e2c7d0eec4ed612e10b94b112dd5f06fcefa994a0c79a45d835cd21eb")
     deprecated("4.3.0", sha256="8e9c3af6dc5a8540b356697081303be392ade3f014615028b3c896d0148397fd")

```

Reply:
> https://github.com/spack/spack/pull/39964 should improve this in the future.

Alright, removed the helper. (I mistakenly pushed the wrong branch in between, sry for the noise)

Reply:
applied now the new context manager